### PR TITLE
Don't lowercase Markdown link definition labels

### PR DESCRIPTION
**Pull request by DerekNonGeneric**
## Description

Markdown link definition label normalization involved case folding
(to lowercase) with undesirable outcomes and has now been modified
to no longer perform case folding of any kind whatever.

Fixes: https://github.com/prettier/prettier/issues/7118
Fixes: https://github.com/prettier/prettier/issues/3835
Signed-off-by: Derek Lewis <DerekNonGeneric@inf.is>

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Clicked the wrong button, was going to run tests. Sorry.

**Comment by DerekNonGeneric** [comment]
@fisker (and everyone else), sorry for the delay on this&hellip; 😅

I think I am about done with this PR now! I am not too familiar with this project's testing strategy and am unsure what else would be expected of me to have this merged. It does indeed fix the problem and will resolve **both** linked issues, but it maybe seems overly trivial to include anything in the changelog regarding it unless I am mistaken.

Does anyone have any thoughts on these changes so far? I'd be glad for any feedback and can always make adjustments if necessary. (I have also allowed maintainers to make changes if so inclined.)

**Comment by DerekNonGeneric** [comment]
@fisker or anyone else (maybe @sosukesuzuki), what else do we need to do for this to be merged? It is a long-standing high-priority reported multiple times and I am personally affected by it almost daily.

Do all of these failing tests showing up in CI need to be modified so that they are passing instead? It seems weird that there are tests here made to ensure the unwanted behavior of casefolding is actually occuring. This fix is not a regression, so unsure on how to proceed. Halp?

**Comment by fisker** [comment]
You need run `yarn test --updateSnapshot`

File: src/language-markdown/clean.js [hunk 1]
```diff
@@ -43,10 +43,7 @@ function clean(ast, newObj, parent) {
   }
 
   if (ast.type === "definition" || ast.type === "linkReference") {
-    newObj.label = ast.label
-      .trim()
-      .replace(/[\t\n ]+/g, " ")
-      .toLowerCase();
+    newObj.label = ast.label.trim().replace(/[\t\n ]+/g, " ");
```
Comment (fisker) [inline src/language-markdown/clean.js:46]: It seem we print `label` as it is, so the value in AST won't change, this line should be removed.

File: tests/format/markdown/linkReference/simple-pascal-case-definition.md [hunk 2]
```diff
@@ -0,0 +1 @@
+[`AsyncGeneratorFunction`]: ./index.html
```
Comment (fisker) [inline tests/format/markdown/linkReference/simple-pascal-case-definition.md:1]: Can you add more tests?

**Comment by fisker** [comment]
It this expected?

**Prettier pr-13155**
[Playground link](https://deploy-preview-13155--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtABOgBgQQM4CeUYA4gnAE4CGMEFAYgK7EwCW0WAukugHQD0rKABM4AD14ALGAFsANiAA0ICAAc20PMlBUKFCAHcACroRaUVOQaoEtygEbUwAazgwAylRlwAMkLjIAGaWeHAOTq4eqlRgQgDmyDAUjGEgcDL2cMKiwj5UUHGMVHFw9HQyNGwFyCBUjLRKINLyAOqSrPB40WBw7mYdrABuHQQ1YHh2IEKhFDBG1HEVQSGpAFZ4Yu7xcnAAiowQ8MtyocrRFDM1FRTOwoZQjaoUQjAtrMIwksgAHAAMZ-pQi1qKoak84DNBgFlABHA7weZqcy1PAAWigcCyWUaFDgcNYuPmxSWSGCJ1SoRkrESyQp2z28ICpJWyhgVHsbw+XyQACZWdRWHJ4gBhCAyElpPAAVkajFCABV2eYyacQIMUgBJEQIDxgZ7qHAidwwAg7Y6hAC+FqAA)

```sh
--parser markdown
```

**Input:**

```markdown
[  `AsyncGeneratorFunction`]: ./index.html
```

**Output:**

```markdown
[  `AsyncGeneratorFunction`]: ./index.html

```

**Comment by DerekNonGeneric** [comment]
> It this expected?

That looks fine to me. Notice how the case is preserved.

As an example of what is currently happening, and
is unacceptable; check out https://github.com/openinf/.github/pull/10/files.

**Comment by fisker** [comment]
> That looks fine to me. 

I don't think so, the spaces should be able to remove.

[`remark-parse` use `collapse-white-space`](https://github.com/remarkjs/remark/blob/daddcb463af2d5b2115496c395d0571c0ff87d15/packages/remark-parse/lib/tokenize/reference.js#L174), let use it too? Just don't lowercase.

**Comment by DerekNonGeneric** [comment]
> let use it too? Just don't lowercase.

Oh, you made that mistake on purpose, but that is somewhat out of scope for this PR. Notice the issues it is supposed to fix.

If you want to re-evaluate how all these link label definitions are handled by prettier at this time, I don't blame you and am willing to participate in that as well.

I agree, that whitespace sh/could have been handled by prettier nicely. 👍

**Comment by fisker** [comment]
> Oh, you made that mistake on purpose

I can't understand.

**Comment by DerekNonGeneric** [comment]
This code sample you provided above for input with the left-padding (`[  `Asy`) is irrelevant to this issue, so not sure why you brought this up. Do you want to see about fixing other problems related to malformed link label definitions? The biggest problem for me is that it becomes lowercased. I would never make the mistake of adding left-padding like your example shows. That is not even a stylist concern as it relates to personal preference it is an obvious mistake made by the author. It could/should be fixed by prettier!

**Comment by fisker** [comment]
Stable version

**Prettier 2.7.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBtABJr2AGBBAZwE8owBxBOAJwEMYIqAxAV1JgEtocBdJdAOgD07KABM4AD34ALGCAA0ICAAcO0AslA0qVCAHcACtoQaUNAG4R2ohSABGtMAGs4MAMo0AtnAAyIuMgAZjQANgRwig40zq5uytEiAObIMFTMESBwnnZwouKiPjRQicw0iXCMDJ50HMXIIDTM9LayniEA6tLs8ATxYHBuJt3s5t1E9WAEGooi4VQwBrSJ1UGh4YoAVgQSbkkhcACKzBDwq2EZ8VRz9dVUTqL6ULbKVCIw7dYw0sgAHAAMiheEHC7VoynqLzgc3MAUUAEdjvBFipTA0CABaKBwXK5WxUOAI9j4xZlFZIYLnRThTzsFJpDIEPaHREBclrDIwGh2D6iL7IABMilSNHYISSAGEIJ4ySAoNBYSBmOEACpc0wU9YgczpACSYgQ7jAr1UeDEbhgRH2Z01WwkACFHC53F5fP5rXAAL4eoA)

```sh
--parser markdown
--arrow-parens avoid
--trailing-comma none
```

**Input:**

```markdown
[          `AsyncGeneratorFunction`]: ./index.ht
```

**Output:**

```markdown
[ `asyncgeneratorfunction`]: ./index.ht

```


This PR

**Prettier pr-13155**
[Playground link](https://deploy-preview-13155--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtABJr2e4AYCCAzgJ5RgDiCcATgIYwQ0BiAruTAJbR4C6S6AHQB6TlAAmcAB6CAFjAC2AGxAAaEBAAOXaEWSg6NGhADuABUMI9KOkpN0Se9QCN6YANZwYAZToK4ADJicMgAZrZEcC5unj6adGBiAObIMDSsUSBwCs5w4pLiAXRQSax0SXDMTAoMXCXIIHSsjGog8soA6rKc8ETxYHDeVj2cAG49JA1gRE4gYpE0MGb0STVhEZkAVkRS3slKcACKrBDw60qR6vE0Cw01NO7iplCtmjRiMB2c4jCyyAAcAAYrsZIh16JoGm84AtRiF1ABHE7wZZaayNIgAWigcDyeVaNDgSM4hOW5TWSHCF0ykQUnFS6Rp+yOyJClI26hgdGcXx+fyQACZOfROEpkgBhCAKClZIgAVlarEiABVudYqZcQKMMgBJCQIHxgd7aAgSbwwEgHc6RAC+NqAA)

```sh
--parser markdown
```

**Input:**

```markdown
[                     `AsyncGeneratorFunction`]: ./index.html
```

**Output:**

```markdown
[                     `AsyncGeneratorFunction`]: ./index.html

```

**Review by fisker — APPROVED** [review]

**Review by DerekNonGeneric — COMMENTED** [review]
Thanks for finishing up here. 👍

**Comment by fisker** [comment]
@thorn0 Can you review?

**Review by duailibe — APPROVED** [review]

**Review by thorn0 — COMMENTED** [review]
looks good

**Comment by DerekNonGeneric** [comment]
thanks, looking forward to the update

**Pull request merged**